### ISSUE TITLE
Can't get inference running on GPU

### ISSUE THREAD

**Issue by TonsOfFun** [ankane/onnxruntime-ruby#9]
First off, great work! I have loaded the `yolov8n.onnx` model from ultralytics and it runs on CPU no problem.

When attempting to run on my GPU I can see the CUDAExecutionProvider is available, but 
```ruby
require "onnxruntime"
require "mini_magick"
require "numo/narray"
OnnxRuntime.ffi_lib = "/app/onnxruntime-linux-x64-gpu-1.15.1/lib/libonnxruntime.so"

session = OnnxRuntime::InferenceSession.new "yolov8n.onnx"
session.providers
```
`=> ["TensorrtExecutionProvider", "CUDAExecutionProvider", "CPUExecutionProvider"]`

but when I run predictions it's only using the CPU

**Comment by ankane** [comment]
Hi @TonsOfFun, I think it's a bug with this library, but not sure when I'll be able to work on it. It looks like `InferenceSession` [needs to call](https://github.com/microsoft/onnxruntime/blob/73037978f85d0a5d4b1f601101dbf65e04a43456/include/onnxruntime/core/session/onnxruntime_c_api.h#L851-L855) `SessionOptionsAppendExecutionProvider_CUDA`.

**Comment by TonsOfFun** [comment]
That's kinda what I was thinking thanks for the suggestion. I'll see if I can get something working and open a PR.

**Comment by ankane** [comment]
Hey @TonsOfFun, fixed in the commit above if you pass `providers: ["CUDAExecutionProvider"]`.

**Comment by TonsOfFun** [comment]
This is awesome! I've been traveling, but I'll test it out tomorrow.

**Comment by ankane** [comment]
It looks like the pre-built GPU version [uses CUDA 11](https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html), so you'll need to either use that or try to compile it from source with CUDA 12.

**Comment by TonsOfFun** [comment]
That was it! I ended up using a modified version of the ml-stack image:
```Dockerfile
FROM nvidia/cuda:11.8.0-cudnn8-devel-ubuntu22.04

ENV RUBY_VERSION 3.1.2

# install packages
RUN apt-get update && \
    DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC apt-get install -y --no-install-recommends \
        autoconf \
        automake \
        build-essential \
        imagemagick \
        libczmq-dev \
        libffi-dev \
        libreadline-dev \
        libsox-dev \
        libsox-fmt-all \
        libssl-dev \
        libtool \
        libvips \
        libyaml-dev \
        libzmq3-dev \
        make \
        python3 \
        python3-pip \
        python3-setuptools \
        sox \
        unzip \
        wget \
        git \
        zlib1g-dev \
        && \
    rm -rf /var/lib/apt/lists/*

# install Ruby
RUN cd /tmp && \
    wget -O ruby.tar.gz -q https://cache.ruby-lang.org/pub/ruby/3.1/ruby-$RUBY_VERSION.tar.gz && \
    mkdir ruby && \
    tar xfz ruby.tar.gz -C ruby --strip-components=1 && \
    rm ruby.tar.gz && \
    cd ruby && \
    ./configure --disable-install-doc --enable-shared && \
    make -j && \
    make install && \
    cd .. && \
    rm -r ruby && \
    ruby --version && \
    bundle --version

# install Jupyter
RUN pip3 install jupyter && \
    jupyter kernelspec remove -f python3 && \
    mkdir /root/.jupyter && \
    echo 'c.KernelSpecManager.ensure_native_kernel = False' > ~/.jupyter/jupyter_notebook_config.py

# install LibTorch
RUN cd /opt && \
    wget -O libtorch.zip -q https://download.pytorch.org/libtorch/cu113/libtorch-cxx11-abi-shared-with-deps-1.12.0%2Bcu113.zip && \
    unzip -q libtorch.zip && \
    rm libtorch.zip

RUN gem install \
        ffi-rzmq \
        iruby \
        mini_magick \
        numo-narray && \
    iruby register

WORKDIR /app

RUN bundle install
# COPY torch-gpu/MNIST.ipynb ./

CMD ["jupyter", "notebook", "--no-browser", "--ip=0.0.0.0", "--allow-root"]
```

With this Gemfile:
```ruby
source "https://rubygems.org"

gem "numo-narray", platform: [:ruby, :x64_mingw]
# until released 0.7.8 is released with GPU support
gem "onnxruntime", git: 'https://github.com/ankane/onnxruntime-ruby.git', ref: 'abf1124' 
gem "mini_magick"
```

In case it's helpful to you or someone else. Thanks again for the support here!
